### PR TITLE
Merge first batch of feature/extensions into main

### PR DESCRIPTION
**Pull request by ryanbogan**
Signed-off-by: Ryan Bogan <rbogan@amazon.com>

### Description
This PR merges the first batch of feature/extensions into main and will then be backported to 2.x.

### Issues Resolved
Issue with links to PR's included in this merge: https://github.com/opensearch-project/OpenSearch/issues/5264
Meta issue for entire feature/extensions merge: https://github.com/opensearch-project/OpenSearch/issues/5263

### Check List
- [X] New functionality includes testing.
  - [X] All tests pass
- [X] New functionality has been documented.
  - [X] New functionality has javadoc added
- [X] Commits are signed per the DCO using --signoff
- [X] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: plugins/discovery-ec2/build.gradle [hunk 1]
```diff
@@ -46,8 +46,6 @@ dependencies {
   api "commons-logging:commons-logging:${versions.commonslogging}"
   api "org.apache.logging.log4j:log4j-1.2-api:${versions.log4j}"
   api "commons-codec:commons-codec:${versions.commonscodec}"
-  api "com.fasterxml.jackson.core:jackson-databind:${versions.jackson_databind}"
```
Comment (dblock) [inline plugins/discovery-ec2/build.gradle:49]: Why? Is this something we can PR separately?

Let's get all the version bumps for dependencies on main first and reduce the size and goal of this PR to extensions only.

File: server/src/main/java/org/opensearch/env/Environment.java [hunk 2]
```diff
@@ -137,6 +139,7 @@ public Environment(final Settings settings, final Path configPath, final boolean
         tmpDir = Objects.requireNonNull(tmpPath);
 
         pluginsDir = homeFile.resolve("plugins");
+        extensionsDir = homeFile.resolve("extensions");
```
Comment (dblock) [inline server/src/main/java/org/opensearch/env/Environment.java:142]: Does this work even if the folder doesn't exist? Since the feature is experimental I wonder whether we need to wrap this into an `if` (I don't think so).

File: server/src/main/java/org/opensearch/extensions/DiscoveryExtension.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+import java.io.IOException;
+import java.util.Map;
+
+/**
+ * Discover extensions running independently or in a separate process
+ *
+ * @opensearch.internal
+ */
+public class DiscoveryExtension extends DiscoveryNode implements Writeable, ToXContentFragment {
```
Comment (dblock) [inline server/src/main/java/org/opensearch/extensions/DiscoveryExtension.java:30]: Should this be called `DiscoveryExtensionNode`? 

It sounds strange either way, are there other names like this? I would expect it to be called `ExtensionNode`. The fact that it was discovered doesn't matter, doesn't it? WDYT?

File: server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
+
+/**
+ * The main class for Plugin Extensibility
+ *
+ * @opensearch.internal
+ */
+public class ExtensionsOrchestrator implements ReportingService<PluginsAndModules> {
```
Comment (dblock) [inline server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:63]: We don't really have other orchestrators in the codebase. Why did we pick this over, say, manager?

File: server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+    List<DiscoveryExtension> extensionsInitializedList;
+    TransportService transportService;
+    ClusterService clusterService;
+
+    public ExtensionsOrchestrator(Settings settings, Path extensionsPath) throws IOException {
+        logger.info("ExtensionsOrchestrator initialized");
+        this.extensionsPath = extensionsPath;
+        this.transportService = null;
+        this.extensionsList = new ArrayList<DiscoveryExtension>();
+        this.extensionsInitializedList = new ArrayList<DiscoveryExtension>();
```
Comment (dblock) [inline server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:98]: I think we end up using initialized extensions by default, so I would flip this to `uninitializedExtensions` and `extensions`.

File: server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+        this.clusterService = null;
+
+        /*
+         * Now Discover extensions
+         */
+        extensionsDiscovery();
```
Comment (dblock) [inline server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:104]: discover()

File: server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java [hunk 7]
```diff
@@ -0,0 +1,386 @@
...
+
+    @Override
+    public PluginsAndModules info() {
+        return null;
+    }
+
+    /*
+     * Load and populate all extensions
+     */
+    private void extensionsDiscovery() throws IOException {
```
Comment (dblock) [inline server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:152]: Rename to `discover()`.

File: server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java [hunk 8]
```diff
@@ -0,0 +1,386 @@
...
+
+        List<Extension> extensions = new ArrayList<Extension>();
+        if (Files.exists(extensionsPath.resolve("extensions.yml"))) {
+            try {
+                extensions = readFromExtensionsYml(extensionsPath.resolve("extensions.yml")).getExtensions();
+            } catch (IOException e) {
+                throw new IOException("Could not read from extensions.yml", e);
+            }
+            for (Extension extension : extensions) {
+                try {
```
Comment (dblock) [inline server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:166]: move to its own private method that loads a single extension

File: server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java [hunk 9]
```diff
@@ -0,0 +1,386 @@
...
+            }
+            if (!extensionsList.isEmpty()) {
+                logger.info("Loaded all extensions");
+            }
+        } else {
+            logger.info("Extensions.yml file is not present.  No extensions will be loaded.");
+        }
+    }
+
+    public void extensionsInitialize() {
```
Comment (dblock) [inline server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:203]: initialize()

File: server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java [hunk 10]
```diff
@@ -0,0 +1,386 @@
...
+        for (DiscoveryNode extensionNode : extensionsList) {
+            extensionInitialize(extensionNode);
+        }
+    }
+
+    private void extensionInitialize(DiscoveryNode extensionNode) {
```
Comment (dblock) [inline server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:209]: initializeExtension

**Review by dblock — CHANGES_REQUESTED** [review]
Glad to see this coming!

I have a bunch of nitpicky comments on naming. I think we should try and get those right since it's all brand new.

I'd like this to come with some .md content in developer guide or somewhere else.

@reta would love your comments

File: plugins/discovery-ec2/build.gradle (hunk 1 above)
Comment (ryanbogan) [reply plugins/discovery-ec2/build.gradle:49]: In order for extensions to be read in, we needed the jackson dependency to be in server.  This caused errors for "unused" dependencies in the plugins folders, since the jackson dependency was already being read in.

File: server/src/main/java/org/opensearch/env/Environment.java (hunk 2 above)
Comment (ryanbogan) [reply server/src/main/java/org/opensearch/env/Environment.java:142]: Yeah it works even if the folder isn't there, but any subsequent actions with that Path won't work without the folder.  I wrapped the areas that use this variable in `if` statements with the feature flag.

File: server/src/main/java/org/opensearch/extensions/DiscoveryExtension.java (hunk 3 above)
Comment (ryanbogan) [reply server/src/main/java/org/opensearch/extensions/DiscoveryExtension.java:30]: I think we should go with `DiscoveryExtensionNode`, since it's extending `DiscoveryNode` but is being tailored for extensions.

File: server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java (hunk 4 above)
Comment (ryanbogan) [reply server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:63]: @owaiskazi19 any thoughts on this?  I wasn't working on extensions when the file was created.

**Comment by ryanbogan** [comment]
@dblock `DEVELOPER_GUIDE.md` of the SDK contains all of the instructions for using extensions, although it will have to be modified with publishing the main branch to mavenLocal and enabling the feature flag.

Here's the file:
https://github.com/opensearch-project/opensearch-sdk-java/blob/main/DEVELOPER_GUIDE.md

File: plugins/discovery-ec2/build.gradle (hunk 1 above)
Comment (ryanbogan) [reply plugins/discovery-ec2/build.gradle:49]: I'll raise a PR separately that adds jackson to server in main and we can merge that one first.

File: server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java (hunk 4 above)
Comment (owaiskazi19) [reply server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:63]: This file is equivalent to what Plugins architecture currently have as [PluginsService.java](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/plugins/PluginsService.java).

**Comment by reta** [comment]
@ryanbogan could you please rebase against `main`? it seems like there are a lot of unrelated changes, thank you

File: server/src/main/java/org/opensearch/common/settings/IndexScopedSettings.java [hunk 11]
```diff
@@ -251,6 +251,11 @@ public IndexScopedSettings copy(Settings settings, IndexMetadata metadata) {
         return new IndexScopedSettings(settings, this, metadata);
     }
 
+    public boolean registerSetting(Setting<?> setting) {
+        validateSettingKey(setting);
```
Comment (reta) [inline server/src/main/java/org/opensearch/common/settings/IndexScopedSettings.java:255]: No need to call `validateSettingsKey`, it is called by `AbstractScopedSettings` already:  https://github.com/opensearch-project/OpenSearch/pull/5347/files#diff-edf64b791aed5932da1f53fa264e42e98bd473f585ceed12892b66cda79c3fd7R154

File: server/src/main/java/org/opensearch/common/settings/SettingsModule.java [hunk 12]
```diff
@@ -179,6 +179,29 @@ public void configure(Binder binder) {
...
 
+    /**
+     * Dynamically registers a new Setting at Runtime. This method is mostly used by plugins/extensions
+     * to register new settings at runtime. Settings can be of Node Scope or Index Scope.
+     * @param setting which is being registered in the cluster.
+     * @return boolean value is set to true when successfully registered, else returns false
+     */
+    public boolean registerDynamicSetting(Setting<?> setting) {
+        try {
+            registerSetting(setting);
```
Comment (reta) [inline server/src/main/java/org/opensearch/common/settings/SettingsModule.java:190]: It seems like we now register settings in several places? What if `clusterSettings.registerSetting` or `indexScopedSettings.registerSetting` fails validation?

File: server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java (hunk 4 above)
Comment (reta) [reply server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:63]: +1 to ExtensionsManager

File: server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java [hunk 13]
```diff
@@ -0,0 +1,394 @@
...
+        REQUEST_EXTENSION_CLUSTER_STATE,
+        REQUEST_EXTENSION_LOCAL_NODE,
+        REQUEST_EXTENSION_CLUSTER_SETTINGS,
+        CREATE_COMPONENT,
+        ON_INDEX_MODULE,
+        GET_SETTINGS
+    };
+
+    private final Path extensionsPath;
+    final List<DiscoveryExtensionNode> uninitializedExtensions;
```
Comment (reta) [inline server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:88]: Please make fields `private`

File: server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java [hunk 14]
```diff
@@ -0,0 +1,394 @@
...
+            false,
+            false,
+            ExtensionRequest::new,
+            ((request, channel, task) -> channel.sendResponse(handleExtensionRequest(request)))
+        );
+    }
+
+    @Override
+    public PluginsAndModules info() {
+        return null;
```
Comment (reta) [inline server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:146]: This should not return `null` I suppose?

File: server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java [hunk 15]
```diff
@@ -0,0 +1,394 @@
...
+                        extension.getJavaVersion(),
+                        extension.getClassName(),
+                        new ArrayList<String>(),
+                        Boolean.parseBoolean(extension.hasNativeController())
+                    )
+                )
+            );
+            logger.info("Loaded extension: " + extension);
+        } catch (IllegalArgumentException e) {
+            logger.error(e.toString());
```
Comment (reta) [inline server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:207]: Shouldn't we fail with `throw new IllegalStateException()` instead of swallowing extension initialization issues?

File: server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java [hunk 16]
```diff
@@ -0,0 +1,394 @@
...
+                    if (extension.getName().equals(response.getName())) {
+                        extensions.add(extension);
+                        break;
+                    }
+                }
+            }
+
+            @Override
+            public void handleException(TransportException exp) {
+                logger.debug(new ParameterizedMessage("Plugin request failed"), exp);
```
Comment (reta) [inline server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:238]: Does it mean extension would stay uninitialized? Should we explicitly write the error message into log?

File: server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java [hunk 17]
```diff
@@ -0,0 +1,394 @@
...
+        try {
+            transportService.connectToNode(extensionNode, true);
+            transportService.sendRequest(
+                extensionNode,
+                REQUEST_EXTENSION_ACTION_NAME,
+                new PluginRequest(transportService.getLocalNode(), new ArrayList<DiscoveryExtensionNode>(uninitializedExtensions)),
+                pluginResponseHandler
+            );
+        } catch (Exception e) {
+            logger.error(e.toString());
```
Comment (reta) [inline server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:255]: Same question as https://github.com/opensearch-project/OpenSearch/pull/5347/files#r1032642750

File: server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java [hunk 18]
```diff
@@ -0,0 +1,394 @@
...
+            );
+            return clusterStateResponse;
+        } else if (extensionRequest.getRequestType() == RequestType.REQUEST_EXTENSION_LOCAL_NODE) {
+            LocalNodeResponse localNodeResponse = new LocalNodeResponse(clusterService);
+            return localNodeResponse;
+        } else if (extensionRequest.getRequestType() == RequestType.REQUEST_EXTENSION_CLUSTER_SETTINGS) {
+            ClusterSettingsResponse clusterSettingsResponse = new ClusterSettingsResponse(clusterService);
+            return clusterSettingsResponse;
+        }
+        throw new Exception("Handler not present for the provided request");
```
Comment (reta) [inline server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:275]:
```suggestion
        throw new Exception("Handler not present for the provided request: " extensionRequest.getRequestType());
```

File: server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java [hunk 19]
```diff
@@ -0,0 +1,394 @@
...
+
+    public void onIndexModule(IndexModule indexModule) throws UnknownHostException {
+        for (DiscoveryNode extensionNode : uninitializedExtensions) {
+            onIndexModule(indexModule, extensionNode);
+        }
+    }
+
+    private void onIndexModule(IndexModule indexModule, DiscoveryNode extensionNode) throws UnknownHostException {
+        logger.info("onIndexModule index:" + indexModule.getIndex());
+        final CountDownLatch inProgressLatch = new CountDownLatch(1);
```
Comment (reta) [inline server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:286]: The usage of `CountDownLatch` seems to be not the best option, there are at least several ways to approach response handlers chaining,  for example using `CompletableFuture<>`s, please let me know if you need help.

**Review by reta — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/opensearch/extensions/ExtensionsSettings.java [hunk 20]
```diff
@@ -0,0 +1,202 @@
...
+
+        private String name;
+        private String uniqueId;
+        private String hostName;
+        private String hostAddress;
+        private String port;
+        private String version;
+        private String description;
+        private String opensearchVersion;
+        private String javaVersion;
```
Comment (reta) [inline server/src/main/java/org/opensearch/extensions/ExtensionsSettings.java:42]: Extension are  out of process, right? So not necessarily JVM based?

Comment (reta) [inline server/src/main/java/org/opensearch/extensions/ExtensionsSettings.java:42]: `javaVersion` -> `jvmVersion` ?

File: server/src/main/java/org/opensearch/index/IndicesModuleNameResponse.java [hunk 21]
```diff
@@ -0,0 +1,42 @@
...
+import org.opensearch.transport.TransportResponse;
+
+import java.io.IOException;
+
+/**
+ * Response for index name of onIndexModule extension point
+ *
+ * @opensearch.internal
+ */
+public class IndicesModuleNameResponse extends TransportResponse {
```
Comment (reta) [inline server/src/main/java/org/opensearch/index/IndicesModuleNameResponse.java:22]: Could we use `AcknowledgedResponse` instead?

File: server/src/main/java/org/opensearch/index/IndicesModuleResponse.java [hunk 22]
```diff
@@ -0,0 +1,85 @@
...
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ * Response for onIndexModule extension point
+ *
+ * @opensearch.internal
+ */
+public class IndicesModuleResponse extends TransportResponse {
+    private boolean addIndexEventListener;
```
Comment (reta) [inline server/src/main/java/org/opensearch/index/IndicesModuleResponse.java:24]: Just an idea, `addIndexEventListener` -> `supportsIndexEventListener` ?

File: server/src/main/java/org/opensearch/indices/IndicesService.java [hunk 23]
```diff
@@ -721,6 +838,9 @@ private synchronized IndexService createIndexService(
             indexModule.addIndexOperationListener(operationListener);
         }
         pluginsService.onIndexModule(indexModule);
+        if (FeatureFlags.isEnabled(FeatureFlags.EXTENSIONS)) {
```
Comment (reta) [inline server/src/main/java/org/opensearch/indices/IndicesService.java:841]: extensionsOrchestrator is `nullable`, better to check it is not `null`

File: server/src/main/java/org/opensearch/node/Node.java [hunk 24]
```diff
@@ -424,6 +426,13 @@ protected Node(
                 initialEnvironment.pluginsDir(),
                 classpathPlugins
             );
+
+            if (FeatureFlags.isEnabled(FeatureFlags.EXTENSIONS)) {
+                this.extensionsOrchestrator = new ExtensionsOrchestrator(tmpSettings, initialEnvironment.extensionDir());
+            } else {
+                this.extensionsOrchestrator = null;
```
Comment (reta) [inline server/src/main/java/org/opensearch/node/Node.java:433]: `Noop` would eliminate dealing with `null` everywhere

File: server/src/main/java/org/opensearch/plugins/PluginsService.java [hunk 25]
```diff
@@ -305,6 +306,9 @@ public Collection<Class<? extends LifecycleComponent>> getGuiceServiceClasses()
     }
 
     public void onIndexModule(IndexModule indexModule) {
+        if (FeatureFlags.isEnabled(FeatureFlags.EXTENSIONS)) {
```
Comment (reta) [inline server/src/main/java/org/opensearch/plugins/PluginsService.java:309]: Why we need that?

File: server/src/main/java/org/opensearch/transport/TransportService.java [hunk 26]
```diff
@@ -397,6 +398,11 @@ public void connectToNode(DiscoveryNode node) throws ConnectTransportException {
         connectToNode(node, (ConnectionProfile) null);
     }
 
+    // We are skipping node validation for extensibility as extensionNode and opensearchNode(LocalNode) will have different ephemeral id's
+    public void connectToNode(final DiscoveryNode node, boolean skipValidation) {
```
Comment (reta) [inline server/src/main/java/org/opensearch/transport/TransportService.java:402]: I would suggest to remove `skipValidation` and introduce dedicated `connectToExtensionNode` method instead

File: server/src/main/java/org/opensearch/transport/TransportService.java [hunk 27]
```diff
@@ -731,6 +775,9 @@ public final <T extends TransportResponse> void sendRequest(
         final TransportResponseHandler<T> handler
     ) {
         try {
+            if (FeatureFlags.isEnabled(FeatureFlags.EXTENSIONS)) {
```
Comment (reta) [inline server/src/main/java/org/opensearch/transport/TransportService.java:778]: Why it is needed?

File: plugins/discovery-ec2/build.gradle (hunk 1 above)
Comment (ryanbogan) [reply plugins/discovery-ec2/build.gradle:49]: I can't figure out why the tests are failing for the PR here: https://github.com/opensearch-project/OpenSearch/actions/runs/3550800781/jobs/5964625474

There is no git diff in the discovery-ec2 `build.gradle` file between this branch and that branch but the tests pass on this branch and fail on that one.

Comment (dblock) [reply plugins/discovery-ec2/build.gradle:49]: Not sure, but raise the PR and we can discuss it there.

File: server/src/main/java/org/opensearch/indices/IndicesService.java (hunk 23 above)
Comment (ryanbogan) [reply server/src/main/java/org/opensearch/indices/IndicesService.java:841]: I think we should keep the FeatureFlag check, because when the feature is moved from experimental to full it will be much easier to keep track of everywhere there are changes and checks.

File: plugins/discovery-ec2/build.gradle (hunk 1 above)
Comment (ryanbogan) [reply plugins/discovery-ec2/build.gradle:49]: https://github.com/opensearch-project/OpenSearch/pull/5366

File: server/src/main/java/org/opensearch/node/Node.java (hunk 24 above)
Comment (ryanbogan) [reply server/src/main/java/org/opensearch/node/Node.java:433]: What would the replacement line be instead of null?

Comment (reta) [reply server/src/main/java/org/opensearch/node/Node.java:433]: Something like `extensionsOrchestrator = new NoopOrchestrator()`

File: server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java (hunk 16 above)
Comment (ryanbogan) [reply server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:238]: The message is written in the `logger.debug`

Comment (reta) [reply server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:238]: By why debug if this is an error? (at least it look like a failure)

Comment (ryanbogan) [reply server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java]: Oh so change it to info instead?

File: server/src/main/java/org/opensearch/common/settings/IndexScopedSettings.java (hunk 11 above)
Comment (ryanbogan) [reply server/src/main/java/org/opensearch/common/settings/IndexScopedSettings.java:255]: The validateSettingsKey method is overridden in `IndexScopedSettings`.  If it's called in the super wouldn't it only call the validateSettingsKey method in `AbstractScopedSettings`?

Comment (reta) [reply server/src/main/java/org/opensearch/common/settings/IndexScopedSettings.java:255]: No, the `validateSettingsKey` will be called for appropriate class: the overridden method in this case.

File: server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java (hunk 19 above)
Comment (ryanbogan) [reply server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java]: I was able to implement half of the functionality with `CompletableFuture<>`, but when I try to invoke the `wait` method to mirror the latch `await` statements that we have, the forbidden API test fails.  Do you know a way other than the `wait` method to set a timeout like the latch does?

File: plugins/repository-azure/build.gradle [hunk 28]
```diff
@@ -44,30 +44,28 @@ opensearchplugin {
 }
 
 dependencies {
-  api 'com.azure:azure-core:1.34.0'
+  api 'com.azure:azure-core:1.33.0'
```
Comment (saratvemulapalli) [inline plugins/repository-azure/build.gradle:47]: Looks like we are going backwards? 
Want to take care of this?

Comment (saratvemulapalli) [reply plugins/repository-azure/build.gradle:47]: Also could you audit bunch of other dependencies which this PR changes?

File: server/src/main/java/org/opensearch/common/settings/AbstractScopedSettings.java [hunk 29]
```diff
@@ -121,8 +122,13 @@ protected AbstractScopedSettings(
                 keySettings.putIfAbsent(setting.getKey(), setting);
             }
         }
-        this.complexMatchers = Collections.unmodifiableMap(complexMatchers);
-        this.keySettings = Collections.unmodifiableMap(keySettings);
+        if (FeatureFlags.isEnabled(FeatureFlags.EXTENSIONS)) {
```
Comment (saratvemulapalli) [inline server/src/main/java/org/opensearch/common/settings/AbstractScopedSettings.java:125]: This is just adding support for dynamically registering settings during runtime. 
It has value for users with/without extensions, WDYT @reta @dblock ?

Comment (reta) [reply server/src/main/java/org/opensearch/common/settings/AbstractScopedSettings.java:125]: @saratvemulapalli good question, I would say if we could split this pull request in two, dynamically registering settings during runtime + extenstion, than the feature flag could be gone for this part.

File: server/src/main/java/org/opensearch/discovery/PluginRequest.java [hunk 30]
```diff
@@ -0,0 +1,76 @@
...
+import org.opensearch.common.io.stream.StreamOutput;
+import org.opensearch.extensions.DiscoveryExtensionNode;
+import org.opensearch.transport.TransportRequest;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * PluginRequest to intialize plugin
```
Comment (saratvemulapalli) [inline server/src/main/java/org/opensearch/discovery/PluginRequest.java:22]: nit: Rename Plugin to Extension? 
```suggestion
 * ExtensionRequest to initialize extension
```

Comment (ryanbogan) [reply server/src/main/java/org/opensearch/discovery/PluginRequest.java:22]: That should be accomplished in the next PR, but I can add it now if you want.

File: server/src/main/java/org/opensearch/common/settings/AbstractScopedSettings.java (hunk 29 above)
Comment (saratvemulapalli) [reply server/src/main/java/org/opensearch/common/settings/AbstractScopedSettings.java:125]: Thanks @reta I can take care of that. Here is the original PR: https://github.com/opensearch-project/OpenSearch/pull/3753

File: server/src/main/java/org/opensearch/discovery/PluginRequest.java (hunk 30 above)
Comment (ryanbogan) [reply server/src/main/java/org/opensearch/discovery/PluginRequest.java:22]: Eventually, it will be InitializeExtensionsRequest based on the current feature/extensions branch

File: server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java (hunk 14 above)
Comment (saratvemulapalli) [reply server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java]: Its mostly how reporting is done. In early days of developing extensions we inherited this from PluginsService.java. 

@ryanbogan can you remove `implements ReportingService<PluginsAndModules>` from `ExtensionsManager` class and remove this `@Override`? 
Also open an issue to track reporting of extensions info and we could make it a simple enhancement.

File: server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java (hunk 16 above)
Comment (saratvemulapalli) [reply server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java]: Yup thats a good point. @ryanbogan lets change it to `logger.error`.

**Comment by saratvemulapalli** [comment]
Thanks @reta @dblock for all the feedback.

File: server/src/main/java/org/opensearch/common/settings/SettingsModule.java (hunk 12 above)
Comment (reta) [reply server/src/main/java/org/opensearch/common/settings/SettingsModule.java:190]: > @reta Thats a good question. Ideally we should undo the changes we made with the call at L190. What do you think of moving this line to L197. This should be safe enough as the validation would throw an error before we register.

@saratvemulapalli this is an option, but in this case we won't register settings in `nodeSettings` and `indexSettings` since the previous statement just return. I was thinking of having the hooks, fe `onNodeSetting`, `onIndexSetting` (or more generic one), where the settings could be added after all validation checks.

**Comment by ryanbogan** [comment]
Precommit is passing locally on my machine

File: server/src/main/java/org/opensearch/common/settings/SettingsModule.java (hunk 12 above)
Comment (ryanbogan) [reply server/src/main/java/org/opensearch/common/settings/SettingsModule.java]: The settings functionality has been removed and added to a PR here: https://github.com/opensearch-project/OpenSearch/pull/5495

**Review by dblock — APPROVED** [review]
Update CHANGELOG and rebase?

**Comment by ryanbogan** [comment]
@reta just need your review and then we can merge or make changes as necessary

**Comment by reta** [comment]
> @reta just need your review and then we can merge or make changes as necessary

I have only concerns regarding this pattern we use in several places 

```
                                /*
                                 * Making async synchronous for now.
                                 */
                                inProgressIndexNameFuture.get(100, TimeUnit.SECONDS);
```

Otherwise LGTM (if you prefer to merge like this, we could address that later by chaining futures instead), thanks!

**Review by reta — APPROVED** [review]

File: plugins/repository-azure/build.gradle [hunk 31]
```diff
@@ -54,7 +54,7 @@ dependencies {
   api "io.netty:netty-resolver-dns:${versions.netty}"
   api "io.netty:netty-transport-native-unix-common:${versions.netty}"
   implementation project(':modules:transport-netty4')
-  api 'com.azure:azure-storage-blob:12.20.0'
+  api 'com.azure:azure-storage-blob:12.20.1'
```
Comment (reta) [inline plugins/repository-azure/build.gradle:57]: It seems to be not related to the change, right?

Comment (ryanbogan) [reply plugins/repository-azure/build.gradle:57]: > > @reta just need your review and then we can merge or make changes as necessary
> 
> I have only concerns regarding this pattern we use in several places
> 
> ```
>                                 /*
>                                  * Making async synchronous for now.
>                                  */
>                                 inProgressIndexNameFuture.get(100, TimeUnit.SECONDS);
> ```
> 
> Otherwise LGTM (if you prefer to merge like this, we could address that later by chaining futures instead), thanks!

There are more latches coming in the next few PR's so let's address that when there are more of them.

Comment (ryanbogan) [reply plugins/repository-azure/build.gradle:57]: Yeah I don't think that's related to this change at all

**Comment by ryanbogan** [comment]
> > @reta just need your review and then we can merge or make changes as necessary
> 
> I have only concerns regarding this pattern we use in several places
> 
> ```
>                                 /*
>                                  * Making async synchronous for now.
>                                  */
>                                 inProgressIndexNameFuture.get(100, TimeUnit.SECONDS);
> ```
> 
> Otherwise LGTM (if you prefer to merge like this, we could address that later by chaining futures instead), thanks!

There are more latches coming in the next few PR's, so let's address that more thoroughly then.

**Comment by saratvemulapalli** [comment]
> > @reta just need your review and then we can merge or make changes as necessary
> 
> I have only concerns regarding this pattern we use in several places
> 
> ```
>                                 /*
>                                  * Making async synchronous for now.
>                                  */
>                                 inProgressIndexNameFuture.get(100, TimeUnit.SECONDS);
> ```
> 
> Otherwise LGTM (if you prefer to merge like this, we could address that later by chaining futures instead), thanks!

Yeah this is because we had to keep the extension points waiting as they are synchronous today. 
@ryanbogan lets open up an issue for all the latches we use and we should move each extension point to have hooks(async).
@reta if you have thoughts on this problem let us know.
